### PR TITLE
Mrc 38 associated companies

### PR DESCRIPTION
**Pull request by adamjw3**

### CODE REVIEW

File: test/controllers/AssociatedCompaniesControllerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    "redirect to Journey Recovery for a POST if no existing data is found" in {
+
+      val application = applicationBuilder(userAnswers = None).build()
+
+      running(application) {
+        val request =
+          FakeRequest(POST, associatedCompaniesRoute)
+            .withFormUrlEncodedBody(("value", AssociatedCompanies.values.head.toString))
```
Comment (girishkamat) [inline test/controllers/AssociatedCompaniesControllerSpec.scala:158]: are you sure this is right? You seem to be sending wrong parameters i.e value isnt supported

File: conf/app.routes [hunk 2]
```diff
@@ -21,6 +21,11 @@ POST       /change-input-screen                 controllers.InputScreenControlle
 
 GET        /results-page                        controllers.ResultsPageController.onPageLoad()
 
+GET        /associated-companies                        controllers.AssociatedCompaniesController.onPageLoad(mode: Mode = NormalMode)
+POST       /associated-companies                        controllers.AssociatedCompaniesController.onSubmit(mode: Mode = NormalMode)
```
Comment (girishkamat) [inline conf/app.routes:25]: can we get the formatting right here

File: conf/messages.en [hunk 3]
```diff
@@ -119,14 +119,22 @@ distribution.error.outOfRange = Distributions must be between {0} and {1}
...
 associatedCompanies.title = Number of associated companies
-associatedCompanies.heading = Number of associated companies
-associatedCompanies.checkYourAnswersLabel = Number of associated companies
-associatedCompanies.error.nonNumeric = Enter your Number of associated companies using numbers
-associatedCompanies.error.required = Enter your Number of associated companies
-associatedCompanies.error.wholeNumber = Enter your Number of associated companies using whole numbers
-associatedCompanies.error.outOfRange = Number of associated companies must be between {0} and {1}
-associatedCompanies.change.hidden = Number of associated companies
-
+associatedCompanies.heading = Did your company have any active associated companies between 1 April 2023 to 31 December 2023?
```
Comment (girishkamat) [inline conf/messages.en:122]: this is supposed to be dynamic right?

File: conf/messages.en [hunk 4]
```diff
@@ -119,14 +119,22 @@ distribution.error.outOfRange = Distributions must be between {0} and {1}
...
+associatedCompanies.warning = Do not include the company you're currently calculating Marginal Relief for in your answer. Doing so will provide you with an incorrect calculation.
+associatedCompanies.detailsTitle = What are associated companies?
+associatedCompanies.detailsHtml = A company is considered associated to another if:<br/><ul><li>one of the companies has <a href="https://www.gov.uk/hmrc-internal-manuals/company-taxation-manual/ctm60210" class="govuk-link">control</a> of the other</li><li>both companies are controlled by the same companies or people</li>
+associatedCompanies.option1Value = yes
```
Comment (girishkamat) [inline conf/messages.en:126]: value should not be in this file, else our backend will fail to work :)

File: test/forms/AssociatedCompaniesFormProviderSpec.scala [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    }
+  }
+
+  private def buildDataMap(associatedCompanies: String, associatedCompaniesCount: String) =
+    Map(
+      s"associatedCompanies"      -> associatedCompanies,
+      s"associatedCompaniesCount" -> associatedCompaniesCount
+    )
+
+  private def buildDataMap2(associatedCompanies: String) =
```
Comment (girishkamat) [inline test/forms/AssociatedCompaniesFormProviderSpec.scala:100]: why do we have this method?

File: test/forms/AssociatedCompaniesFormProviderSpec.scala [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+  "form values" - {
+
+    "Are valid" in {
+      val range = intsInRangeWithCommas(0, 99)
+
+      forAll(range) { range =>
+        val data = buildDataMap("yes", range.toString())
+        val result = form.bind(data)
+        result.hasErrors mustBe false
```
Comment (girishkamat) [inline test/forms/AssociatedCompaniesFormProviderSpec.scala:37]: also worth comparing the form value to the expected result

File: test/controllers/AssociatedCompaniesControllerSpec.scala (hunk 1 above)
Comment (adamjw3) [reply test/controllers/AssociatedCompaniesControllerSpec.scala:158]: fixed

File: conf/app.routes (hunk 2 above)
Comment (adamjw3) [reply conf/app.routes:25]: fixed

File: conf/messages.en [hunk 7]
```diff
@@ -119,14 +119,22 @@ distribution.error.outOfRange = Distributions must be between {0} and {1}
...
+associatedCompanies.option1Label = Yes
+associatedCompanies.option2Value = no
+associatedCompanies.option2Label = No
+associatedCompanies.countLabel = Enter only additional associated companies
+associatedCompanies.checkYourAnswersLabel = Did your company have any active associated companies between 1 April 2023 to 31 December 2023?
+associatedCompanies.error.required = Select whether you have associated companies or not
+associatedCompaniesCount.error.wholeNumber = Do not use decimal points. Enter a whole number. For example, 1
+associatedCompaniesCount.error.nonNumeric = Enter your Number of associated companies using numbers
+associatedCompaniesCount.error.outOfRange = Enter an amount that is less than or equal to 99
```
[the remark this answers is not in the document]
Comment (adamjw3) [reply conf/messages.en:135]: fixed

File: conf/messages.en (hunk 4 above)
Comment (adamjw3) [reply conf/messages.en]: Moved option values and text inline in the view

File: test/forms/AssociatedCompaniesFormProviderSpec.scala (hunk 5 above)
Comment (girishkamat) [reply test/forms/AssociatedCompaniesFormProviderSpec.scala:102]: You could combine the 2 methods into 1

```
  private def buildDataMap(associatedCompanies: String, associatedCompaniesCount: Option[String] = None) =
    Map(
      s"associatedCompanies"      -> associatedCompanies,
    ) ++ associatedCompaniesCount.map(c => "associatedCompaniesCount" -> c)
```

**Pull request merged**